File: src/stores/plugins.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { invoke } from '@tauri-apps/api/core'

export type PluginPermission = 'storage' | 'network' | 'filesystem' | 'notifications' | 'tabs' | 'bookmarks' | 'history' | 'downloads'
export type PluginHook = 'page_load' | 'tab_created' | 'tab_closed' | 'bookmark_added' | 'download_started' | 'url_changed'

export interface PluginSetting {
  key: string
  value: any
  type: 'string' | 'number' | 'boolean' | 'array' | 'object'
  description?: string
}

export interface PluginManifest {
  name: string
  version: string
  description: string
  author: string
  permissions: PluginPermission[]
  hooks: PluginHook[]
  settings: PluginSetting[]
  icon?: string
  homepage?: string
}

export interface Plugin {
  id: string
  manifest: PluginManifest
  enabled: boolean
  installed_at: string
  last_updated: string
  file_path: string
  settings: Record<string, any>
}

export interface PluginEvent {
  hook: PluginHook
  data: any
  timestamp: string
}

export interface PluginStats {
  total_plugins: number
  enabled_plugins: number
  disabled_plugins: number
  events_processed: number
  last_event_time?: string
}

export const usePluginsStore = defineStore('plugins', () => {
  const plugins = ref<Plugin[]>([])
  const isLoading = ref(false)

  async function loadPlugins(): Promise<void> {
    try {
      isLoading.value = true
      const pluginList = await invoke<Plugin[]>('get_all_plugins')
      plugins.value = pluginList
    } catch (error) {
      console.error('Failed to load plugins:', error)
      throw error
    } finally {
      isLoading.value = false
    }
  }

  async function installPlugin(filePath: string): Promise<string> {
    try {
      isLoading.value = true
      const pluginId = await invoke<string>('install_plugin', { filePath })
      await loadPlugins()
      return pluginId
    } catch (error) {
      console.error('Failed to install plugin:', error)
      throw error
    } finally {
      isLoading.value = false
    }
  }

  async function uninstallPlugin(pluginId: string): Promise<void> {
    try {
      isLoading.value = true
      await invoke('uninstall_plugin', { pluginId })
      plugins.value = plugins.value.filter(plugin => plugin.id !== pluginId)
    } catch (error) {
      console.error('Failed to uninstall plugin:', error)
      throw error
    } finally {
      isLoading.value = false
    }
  }

  async function enablePlugin(pluginId: string): Promise<void> {
    try {
      await invoke('enable_plugin', { pluginId })
      const plugin = plugins.value.find(p => p.id === pluginId)
      if (plugin) {
        plugin.enabled = true
      }
    } catch (error) {
      console.error('Failed to enable plugin:', error)
      throw error
    }
  }

  async function disablePlugin(pluginId: string): Promise<void> {
    try {
      await invoke('disable_plugin', { pluginId })
      const plugin = plugins.value.find(p => p.id === pluginId)
      if (plugin) {
        plugin.enabled = false
      }
    } catch (error) {
      console.error('Failed to disable plugin:', error)
      throw error
    }
  }

  async function getPlugin(pluginId: string): Promise<Plugin | null> {
    try {
      const plugin = await invoke<Plugin>('get_plugin', { pluginId })
      return plugin
    } catch (error) {
      console.error('Failed to get plugin:', error)
      return null
    }
  }

  async function getEnabledPlugins(): Promise<Plugin[]> {
    try {
      const enabledPlugins = await invoke<Plugin[]>('get_enabled_plugins')
      return enabledPlugins
    } catch (error) {
      console.error('Failed to get enabled plugins:', error)
      return []
    }
  }

  async function updatePluginSetting(pluginId: string, key: string, value: any): Promise<void> {
    try {
      await invoke('update_plugin_setting', { pluginId, key, value })
      const plugin = plugins.value.find(p => p.id === pluginId)
      if (plugin) {
        plugin.settings[key] = value
      }
    } catch (error) {
      console.error('Failed to update plugin setting:', error)
      throw error
    }
  }

  async function fetchPluginSetting(pluginId: string, key: string): Promise<any> {
    try {
      const value = await invoke('get_plugin_setting', { pluginId, key })
      return value
    } catch (error) {
      console.error('Failed to get plugin setting:', error)
      return null
    }
  }

  async function triggerPluginEvent(hook: PluginHook, data: any): Promise<void> {
    try {
      await invoke('trigger_plugin_event', { hook, data })
    } catch (error) {
      console.error('Failed to trigger plugin event:', error)
      throw error
    }
  }

  async function hasPluginPermission(pluginId: string, permission: PluginPermission): Promise<boolean> {
    try {
      const hasPermission = await invoke<boolean>('has_plugin_permission', { pluginId, permission })
      return hasPermission
    } catch (error) {
      console.error('Failed to check plugin permission:', error)
      return false
    }
  }

  async function getPluginsByHook(hook: PluginHook): Promise<Plugin[]> {
    try {
      const hookPlugins = await invoke<Plugin[]>('get_plugins_by_hook', { hook })
      return hookPlugins
    } catch (error) {
      console.error('Failed to get plugins by hook:', error)
      return []
    }
  }

  async function searchPlugins(query: string): Promise<Plugin[]> {
    try {
      const results = await invoke<Plugin[]>('search_plugins', { query })
      return results
    } catch (error) {
      console.error('Failed to search plugins:', error)
      return []
    }
  }

  async function exportPluginSettings(): Promise<string> {
    try {
      const data = await invoke<string>('export_plugin_settings')
      return data
    } catch (error) {
      console.error('Failed to export plugin settings:', error)
      throw error
    }
  }

  async function importPluginSettings(data: string): Promise<void> {
    try {
      isLoading.value = true
      await invoke('import_plugin_settings', { data })
      await loadPlugins()
    } catch (error) {
      console.error('Failed to import plugin settings:', error)
      throw error
    } finally {
      isLoading.value = false
    }
  }

  async function getPluginStats(): Promise<PluginStats> {
    try {
      const stats = await invoke<PluginStats>('get_plugin_stats')
      return stats
    } catch (error) {
      console.error('Failed to get plugin stats:', error)
      return {
        total_plugins: 0,
        enabled_plugins: 0,
        disabled_plugins: 0,
        events_processed: 0
      }
    }
  }

  async function validatePluginManifest(manifestPath: string): Promise<{ valid: boolean; errors: string[] }> {
    try {
      const result = await invoke<{ valid: boolean; errors: string[] }>('validate_plugin_manifest', { manifestPath })
      return result
    } catch (error) {
      console.error('Failed to validate plugin manifest:', error)
      return { valid: false, errors: ['Failed to validate manifest'] }
    }
  }

  function getPluginById(pluginId: string): Plugin | undefined {
    return plugins.value.find(plugin => plugin.id === pluginId)
  }

  function getEnabledPluginsSync(): Plugin[] {
    return plugins.value.filter(plugin => plugin.enabled)
  }

  function getDisabledPlugins(): Plugin[] {
    return plugins.value.filter(plugin => !plugin.enabled)
  }

  function getPluginsByPermission(permission: PluginPermission): Plugin[] {
    return plugins.value.filter(plugin => 
      plugin.manifest.permissions.includes(permission)
    )
  }

  function getPluginsByHookSync(hook: PluginHook): Plugin[] {
    return plugins.value.filter(plugin => 
      plugin.enabled && plugin.manifest.hooks.includes(hook)
    )
  }

  function hasPermission(plugin: Plugin, permission: PluginPermission): boolean {
    return plugin.manifest.permissions.includes(permission)
  }

  function hasHook(plugin: Plugin, hook: PluginHook): boolean {
    return plugin.manifest.hooks.includes(hook)
  }

  function getPluginSettingSync(plugin: Plugin, key: string): any {
    return plugin.settings[key]
  }

  function getPluginSettingWithDefault(plugin: Plugin, key: string, defaultValue: any): any {
    const setting = plugin.manifest.settings.find(s => s.key === key)
    if (setting && plugin.settings[key] !== undefined) {
      return plugin.settings[key]
    }
    return setting?.value ?? defaultValue
  }

  function formatPluginSize(_: Plugin): string {
    return 'Unknown'
  }

  function getPluginVersion(plugin: Plugin): string {
    return plugin.manifest.version
  }

  function isPluginOutdated(_: Plugin): boolean {
    return false
  }

  function getPluginDependencies(_: Plugin): string[] {
    return []
  }

  return {
    plugins,
    isLoading,
    loadPlugins,
    installPlugin,
    uninstallPlugin,
    enablePlugin,
    disablePlugin,
    getPlugin,
    getEnabledPlugins,
    updatePluginSetting,
    fetchPluginSetting,
    triggerPluginEvent,
    hasPluginPermission,
    getPluginsByHook,
    searchPlugins,
    exportPluginSettings,
    importPluginSettings,
    getPluginStats,
    validatePluginManifest,
    getPluginById,
    getEnabledPluginsSync,
    getDisabledPlugins,
    getPluginsByPermission,
    getPluginsByHookSync,
    hasPermission,
    hasHook,
    getPluginSettingSync,
    getPluginSettingWithDefault,
    formatPluginSize,
    getPluginVersion,
    isPluginOutdated,
    getPluginDependencies
  }
})

File: src-tauri/src/browser/plugins.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tokio::sync::RwLock;
use once_cell::sync::Lazy;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Plugin {
    pub id: String,
    pub name: String,
    pub version: String,
    pub description: String,
    pub author: String,
    pub enabled: bool,
    pub permissions: Vec<PluginPermission>,
    pub hooks: Vec<PluginHook>,
    pub settings: HashMap<String, PluginSetting>,
    pub manifest_path: String,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum PluginPermission {
    NetworkAccess,
    FileSystemRead,
    FileSystemWrite,
    BookmarkAccess,
    HistoryAccess,
    TabManagement,
    WindowManagement,
    SettingsAccess,
    NotificationAccess,
    ClipboardAccess,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub enum PluginHook {
    BeforeNavigate,
    AfterNavigate,
    BeforeRequest,
    AfterRequest,
    TabCreated,
    TabClosed,
    WindowCreated,
    WindowClosed,
    BookmarkAdded,
    BookmarkRemoved,
    HistoryAdded,
    DownloadStarted,
    DownloadCompleted,
    SettingsChanged,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PluginSetting {
    pub key: String,
    pub value: serde_json::Value,
    pub setting_type: PluginSettingType,
    pub description: String,
    pub required: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum PluginSettingType {
    String,
    Number,
    Boolean,
    Array,
    Object,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PluginManifest {
    pub name: String,
    pub version: String,
    pub description: String,
    pub author: String,
    pub main: String,
    pub permissions: Vec<PluginPermission>,
    pub hooks: Vec<PluginHook>,
    pub settings: Vec<PluginSettingDefinition>,
    pub min_browser_version: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PluginSettingDefinition {
    pub key: String,
    pub setting_type: PluginSettingType,
    pub description: String,
    pub default_value: serde_json::Value,
    pub required: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PluginEvent {
    pub plugin_id: String,
    pub hook: PluginHook,
    pub data: serde_json::Value,
    pub timestamp: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PluginStats {
    pub total_plugins: usize,
    pub enabled_plugins: usize,
    pub disabled_plugins: usize,
    pub events_processed: u64,
    pub last_event_time: Option<chrono::DateTime<chrono::Utc>>,
}

static PLUGIN_MANAGER: Lazy<RwLock<PluginManager>> = Lazy::new(|| {
    RwLock::new(PluginManager::new())
});

pub struct PluginManager {
    pub plugins: HashMap<String, Plugin>,
    pub event_handlers: HashMap<PluginHook, Vec<String>>,
    pub stats: PluginStats,
}

impl PluginManager {
    pub fn new() -> Self {
        Self {
            plugins: HashMap::new(),
            event_handlers: HashMap::new(),
            stats: PluginStats {
                total_plugins: 0,
                enabled_plugins: 0,
                disabled_plugins: 0,
                events_processed: 0,
                last_event_time: None,
            },
        }
    }

    pub fn install_plugin(&mut self, manifest_path: &str) -> Result<String, String> {
        let manifest_content = std::fs::read_to_string(manifest_path)
            .map_err(|e| format!("Failed to read manifest: {}", e))?;
        
        let manifest: PluginManifest = serde_json::from_str(&manifest_content)
            .map_err(|e| format!("Invalid manifest format: {}", e))?;
        
        let plugin_id = Uuid::new_v4().to_string();
        let mut settings = HashMap::new();
        
        for setting_def in &manifest.settings {
            settings.insert(setting_def.key.clone(), PluginSetting {
                key: setting_def.key.clone(),
                value: setting_def.default_value.clone(),
                setting_type: setting_def.setting_type.clone(),
                description: setting_def.description.clone(),
                required: setting_def.required,
            });
        }
        
        let plugin = Plugin {
            id: plugin_id.clone(),
            name: manifest.name,
            version: manifest.version,
            description: manifest.description,
            author: manifest.author,
            enabled: true,
            permissions: manifest.permissions,
            hooks: manifest.hooks.clone(),
            settings,
            manifest_path: manifest_path.to_string(),
            created_at: chrono::Utc::now(),
            updated_at: chrono::Utc::now(),
        };
        
        for hook in &manifest.hooks {
            self.event_handlers.entry(hook.clone())
                .or_insert_with(Vec::new)
                .push(plugin_id.clone());
        }
        
        self.plugins.insert(plugin_id.clone(), plugin);
        self.update_stats();
        
        Ok(plugin_id)
    }

    pub fn uninstall_plugin(&mut self, plugin_id: &str) -> Result<(), String> {
        let plugin = self.plugins.remove(plugin_id)
            .ok_or("Plugin not found")?;
        
        for hook in &plugin.hooks {
            if let Some(handlers) = self.event_handlers.get_mut(hook) {
                handlers.retain(|id| id != plugin_id);
                if handlers.is_empty() {
                    self.event_handlers.remove(hook);
                }
            }
        }
        
        self.update_stats();
        Ok(())
    }

    pub fn enable_plugin(&mut self, plugin_id: &str) -> Result<(), String> {
        let plugin = self.plugins.get_mut(plugin_id)
            .ok_or("Plugin not found")?;
        
        plugin.enabled = true;
        plugin.updated_at = chrono::Utc::now();
        
        for hook in &plugin.hooks {
            self.event_handlers.entry(hook.clone())
                .or_insert_with(Vec::new)
                .push(plugin_id.to_string());
        }
        
        self.update_stats();
        Ok(())
    }

    pub fn disable_plugin(&mut self, plugin_id: &str) -> Result<(), String> {
        let plugin = self.plugins.get_mut(plugin_id)
            .ok_or("Plugin not found")?;
        
        plugin.enabled = false;
        plugin.updated_at = chrono::Utc::now();
        
        for hook in &plugin.hooks {
            if let Some(handlers) = self.event_handlers.get_mut(hook) {
                handlers.retain(|id| id != plugin_id);
                if handlers.is_empty() {
                    self.event_handlers.remove(hook);
                }
            }
        }
        
        self.update_stats();
        Ok(())
    }

    pub fn get_plugin(&self, plugin_id: &str) -> Option<&Plugin> {
        self.plugins.get(plugin_id)
    }

    pub fn get_all_plugins(&self) -> Vec<&Plugin> {
        self.plugins.values().collect()
    }

    pub fn get_enabled_plugins(&self) -> Vec<&Plugin> {
        self.plugins.values().filter(|p| p.enabled).collect()
    }

    pub fn update_plugin_setting(&mut self, plugin_id: &str, key: &str, value: serde_json::Value) -> Result<(), String> {
        let plugin = self.plugins.get_mut(plugin_id)
            .ok_or("Plugin not found")?;
        
        if let Some(setting) = plugin.settings.get_mut(key) {
            setting.value = value;
            plugin.updated_at = chrono::Utc::now();
            Ok(())
        } else {
            Err("Setting not found".to_string())
        }
    }

    pub fn get_plugin_setting(&self, plugin_id: &str, key: &str) -> Option<&PluginSetting> {
        self.plugins.get(plugin_id)?
            .settings.get(key)
    }

    pub fn trigger_event(&mut self, hook: PluginHook, data: serde_json::Value) -> Vec<String> {
        let mut triggered_plugins = Vec::new();
        
        if let Some(handlers) = self.event_handlers.get(&hook) {
            for plugin_id in handlers {
                if let Some(plugin) = self.plugins.get(plugin_id) {
                    if plugin.enabled {
                        triggered_plugins.push(plugin_id.clone());
                    }
                }
            }
        }
        
        self.stats.events_processed += 1;
        self.stats.last_event_time = Some(chrono::Utc::now());
        
        triggered_plugins
    }

    pub fn has_permission(&self, plugin_id: &str, permission: &PluginPermission) -> bool {
        if let Some(plugin) = self.plugins.get(plugin_id) {
            plugin.enabled && plugin.permissions.contains(permission)
        } else {
            false
        }
    }

    pub fn get_plugins_by_hook(&self, hook: &PluginHook) -> Vec<&Plugin> {
        self.plugins.values()
            .filter(|p| p.enabled && p.hooks.contains(hook))
            .collect()
    }

    pub fn search_plugins(&self, query: &str) -> Vec<&Plugin> {
        let query_lower = query.to_lowercase();
        self.plugins.values()
            .filter(|p| {
                p.name.to_lowercase().contains(&query_lower) ||
                p.description.to_lowercase().contains(&query_lower) ||
                p.author.to_lowercase().contains(&query_lower)
            })
            .collect()
    }

    pub fn export_plugin_settings(&self, plugin_id: &str) -> Result<String, String> {
        let plugin = self.plugins.get(plugin_id)
            .ok_or("Plugin not found")?;
        
        serde_json::to_string_pretty(&plugin.settings)
            .map_err(|e| format!("Failed to export settings: {}", e))
    }

    pub fn import_plugin_settings(&mut self, plugin_id: &str, settings_data: &str) -> Result<(), String> {
        let settings: HashMap<String, PluginSetting> = serde_json::from_str(settings_data)
            .map_err(|e| format!("Invalid settings format: {}", e))?;
        
        let plugin = self.plugins.get_mut(plugin_id)
            .ok_or("Plugin not found")?;
        
        for (key, setting) in settings {
            if plugin.settings.contains_key(&key) {
                plugin.settings.insert(key, setting);
            }
        }
        
        plugin.updated_at = chrono::Utc::now();
        Ok(())
    }

    pub fn get_stats(&self) -> &PluginStats {
        &self.stats
    }

    fn update_stats(&mut self) {
        self.stats.total_plugins = self.plugins.len();
        self.stats.enabled_plugins = self.plugins.values().filter(|p| p.enabled).count();
        self.stats.disabled_plugins = self.stats.total_plugins - self.stats.enabled_plugins;
    }

    pub fn validate_plugin_manifest(&self, manifest_path: &str) -> Result<PluginManifest, String> {
        let manifest_content = std::fs::read_to_string(manifest_path)
            .map_err(|e| format!("Failed to read manifest: {}", e))?;
        
        let manifest: PluginManifest = serde_json::from_str(&manifest_content)
            .map_err(|e| format!("Invalid manifest format: {}", e))?;
        
        if manifest.name.is_empty() {
            return Err("Plugin name cannot be empty".to_string());
        }
        
        if manifest.version.is_empty() {
            return Err("Plugin version cannot be empty".to_string());
        }
        
        Ok(manifest)
    }
}

#[tauri::command]
pub async fn install_plugin(manifest_path: String) -> Result<String, String> {
    let mut manager = PLUGIN_MANAGER.write().await;
    manager.install_plugin(&manifest_path)
}

#[tauri::command]
pub async fn uninstall_plugin(plugin_id: String) -> Result<(), String> {
    let mut manager = PLUGIN_MANAGER.write().await;
    manager.uninstall_plugin(&plugin_id)
}

#[tauri::command]
pub async fn enable_plugin(plugin_id: String) -> Result<(), String> {
    let mut manager = PLUGIN_MANAGER.write().await;
    manager.enable_plugin(&plugin_id)
}

#[tauri::command]
pub async fn disable_plugin(plugin_id: String) -> Result<(), String> {
    let mut manager = PLUGIN_MANAGER.write().await;
    manager.disable_plugin(&plugin_id)
}

#[tauri::command]
pub async fn get_plugin(plugin_id: String) -> Result<Option<Plugin>, String> {
    let manager = PLUGIN_MANAGER.read().await;
    Ok(manager.get_plugin(&plugin_id).cloned())
}

#[tauri::command]
pub async fn get_all_plugins() -> Result<Vec<Plugin>, String> {
    let manager = PLUGIN_MANAGER.read().await;
    Ok(manager.get_all_plugins().into_iter().cloned().collect())
}

#[tauri::command]
pub async fn get_enabled_plugins() -> Result<Vec<Plugin>, String> {
    let manager = PLUGIN_MANAGER.read().await;
    Ok(manager.get_enabled_plugins().into_iter().cloned().collect())
}

#[tauri::command]
pub async fn update_plugin_setting(plugin_id: String, key: String, value: serde_json::Value) -> Result<(), String> {
    let mut manager = PLUGIN_MANAGER.write().await;
    manager.update_plugin_setting(&plugin_id, &key, value)
}

#[tauri::command]
pub async fn get_plugin_setting(plugin_id: String, key: String) -> Result<Option<PluginSetting>, String> {
    let manager = PLUGIN_MANAGER.read().await;
    Ok(manager.get_plugin_setting(&plugin_id, &key).cloned())
}

#[tauri::command]
pub async fn trigger_plugin_event(hook: PluginHook, data: serde_json::Value) -> Result<Vec<String>, String> {
    let mut manager = PLUGIN_MANAGER.write().await;
    Ok(manager.trigger_event(hook, data))
}

#[tauri::command]
pub async fn has_plugin_permission(plugin_id: String, permission: PluginPermission) -> Result<bool, String> {
    let manager = PLUGIN_MANAGER.read().await;
    Ok(manager.has_permission(&plugin_id, &permission))
}

#[tauri::command]
pub async fn get_plugins_by_hook(hook: PluginHook) -> Result<Vec<Plugin>, String> {
    let manager = PLUGIN_MANAGER.read().await;
    Ok(manager.get_plugins_by_hook(&hook).into_iter().cloned().collect())
}

#[tauri::command]
pub async fn search_plugins(query: String) -> Result<Vec<Plugin>, String> {
    let manager = PLUGIN_MANAGER.read().await;
    Ok(manager.search_plugins(&query).into_iter().cloned().collect())
}

#[tauri::command]
pub async fn export_plugin_settings(plugin_id: String) -> Result<String, String> {
    let manager = PLUGIN_MANAGER.read().await;
    manager.export_plugin_settings(&plugin_id)
}

#[tauri::command]
pub async fn import_plugin_settings(plugin_id: String, settings_data: String) -> Result<(), String> {
    let mut manager = PLUGIN_MANAGER.write().await;
    manager.import_plugin_settings(&plugin_id, &settings_data)
}

#[tauri::command]
pub async fn get_plugin_stats() -> Result<PluginStats, String> {
    let manager = PLUGIN_MANAGER.read().await;
    Ok(manager.get_stats().clone())
}

#[tauri::command]
pub async fn validate_plugin_manifest(manifest_path: String) -> Result<PluginManifest, String> {
    let manager = PLUGIN_MANAGER.read().await;
    manager.validate_plugin_manifest(&manifest_path)
}